core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -146,24 +152,32 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
 
                                final long nextDelay = getNextDelay(ctx, backoff, millisAfter);
                                if (nextDelay < 0) {
-                                   finishRetryWithCurrentResponse(ctx, rootReqDuplicator, res, resDuplicator);
+                                   finishRetryWithCurrentResponse(
+                                           ctx, rootReqDuplicator, future, resDuplicator);
                                    return;
                                }
 
                                resDuplicator.close();
-                               final Consumer<? super Throwable> actionOnException = cause -> {
-                                   onRetryingComplete(ctx);
-                                   res.completeExceptionally(cause);
-                                   rootReqDuplicator.close();
-                               };
-                               final Runnable retryTask = () -> doExecute0(ctx, rootReqDuplicator, res);
-                               scheduleNextRetry(ctx, actionOnException, retryTask, nextDelay);
+                               scheduleNextRetry(ctx, actionOnException(ctx, rootReqDuplicator, future),
+                                                 () -> doExecute0(ctx, rootReqDuplicator, originalReq,
+                                                                  returnedRes, future),
+                                                 nextDelay);
                            } else {
-                               finishRetryWithCurrentResponse(ctx, rootReqDuplicator, res, resDuplicator);
+                               finishRetryWithCurrentResponse(ctx, rootReqDuplicator, future, resDuplicator);
                            }
                        });
     }
 
+    private static Consumer<? super Throwable> actionOnException(ClientRequestContext ctx,
+                                                                 HttpRequestDuplicator rootReqDuplicator,
+                                                                 CompletableFuture<HttpResponse> future) {
+        return cause -> {
+            onRetryingComplete(ctx);
+            future.completeExceptionally(cause);
+            rootReqDuplicator.close();
```

Comment:
How about extracting this into a method like `handleException()` so that we don't need to do `actionOnException().accept()`?